src/FunctionApps/python/IntakePipeline/conversion.py

```diff
@@ -1,18 +1,130 @@
 import logging
 import re
 import requests
+import hl7
 from typing import List, Dict
 
 
-def clean_message(message: str, delimiter: str = "\n") -> str:
-    cleaned_message = re.sub("[\r\n]+", delimiter, message)
+def clean_message(message: str) -> str:
+    """Prepare a message for conversion by adjusting problematic fields 
+    to conform to Azure's expectations.
+    * Convert segment terminators from \n to \r
+    * Normalize datetime fields 
+    """
+    parsed_message = ""
+    try:
+        # Conversion from \n to \r EOL characters is needed for hl7 
+        # module, and doesn't hurt the Azure converter
+        parsed_message = hl7.parse(message.replace("\n","\r"))
+
+        # Normalize Dates
+        # MSH-7 - Message date/time
+        normalize_hl7_datetime_segment(parsed_message, "MSH", [7])
+
+        # PID-7 - DOB
+        normalize_hl7_datetime_segment(parsed_message, "PID", [7])
+
+        # PV1-44 - Admisstion Date
+        # PV1-45 - Discharge Date
+        normalize_hl7_datetime_segment(parsed_message, "PV1", [44,45])
+
+        # ORC-9 Date/time of transaction
+        # ORC-15 Order effective date/time
+        # ORC-27 Filler's expected availability date/time
+        normalize_hl7_datetime_segment(parsed_message, "ORC", [9,15,27])
+
+        # OBR-7 Observation date/time
+        # OBR-8 Observation end date/time
+        # OBR-22 Status change date/time
+        # OBR-36 Scheduled date/time
+        normalize_hl7_datetime_segment(parsed_message, "OBR", [7,8,22,36]) 
+
+        # OBX-12 Effective date/time of reference range
+        # OBX-14 Date/time of observation
+        # OBX-19 Date/time of analysis
+        normalize_hl7_datetime_segment(parsed_message, "OBX", [12,14,19])
+        
+        # TQ1-7 Start date/time
+        # TQ1-8 End date/time
+        normalize_hl7_datetime_segment(parsed_message, "TQ1", [7,8])
+        
+        # SPM-18 Specimen received date/time
+        # SPM-19 Specimen expiration date/time
+        normalize_hl7_datetime_segment(parsed_message, "SPM", [18,19])
+
+        # RXA-3 Date/time start of administration
+        # RXA-4 Date/time end of administration
+        # RXA-16 Substance expiration date
+        # RXA-22 System entry date/time
+        normalize_hl7_datetime_segment(parsed_message, "RXA", [3,4,16,22])
```

Comment:
Could we condense all of the calls of normalize_hl7_datetime_segement to looping over a dictionary with key:value pairs of the form hl7_segment:list of indexes? Might make it easier to update the places we look for dates in the future. Also possibly to support adapting to other  HL7 formats. This is clean and well documented as is though.